git/checkignore.go

```diff
@@ -0,0 +1,244 @@
+
+		for {
+			gitignore := filepath.Join(dir, ".gitignore")
+			log.Printf("Checking .gitignore in %s\n", gitignore)
+			ignorePath, err := filepath.Rel(dir, abs)
+			pattern, lineNumber, err := findPatternInIgnoreFile(c, gitignore, ignorePath, isDir)
+			if err != nil {
+				return nil, err
+			}
+
+			if pattern != "" {
+				gitignore, _ = filepath.Rel(c.WorkDir.String(), gitignore)
+				patternMatches[idx] = IgnoreMatch{Pattern: pattern, LineNum: lineNumber, Source: File(gitignore), PathName: path}
+
+				break
+			}
+
+			if dir == c.WorkDir.String() {
+				break
+			}
+			dir = filepath.Dir(dir)
+		}
+
+		// Check .git/info/exclude
+		if patternMatches[idx].Pattern == "" {
+			pattern, lineNumber, err := findPatternInIgnoreFile(c, filepath.Join(c.GitDir.String(), "info/exclude"), wdpath, isDir)
+			if err != nil {
+				return nil, err
+			}
+
+			if pattern != "" {
+				patternMatches[idx] = IgnoreMatch{Pattern: pattern, LineNum: lineNumber, Source: File(".git/info/exclude"), PathName: path} // TODO .git/info/exclude is hard coded here instead of being calculated from c.GitDir
+			}
+		}
+
+		// Be sure to assign the pathname in all cases so that clients can match the outputs to their inputs
+		patternMatches[idx].PathName = path
+
+		// TODO: consider the other places where ignores can come from, such as core.excludesFile and .git/info/exclude
+	}
+
+	return patternMatches, nil
+}
+
+// Finds the first matching pattern and line number in the provided ignoreFile that matches the ignorePath, relative to the
+//  the ignoreFile, if one exists. If no pattern matches or the ignoreFile doesn't exist then the returned pattern is an empty string.
+//  If the ignoreFile is not within the work directory then the ignorePath should be made relative to the work directory.
+func findPatternInIgnoreFile(c *Client, ignoreFile string, ignorePath string, isDir bool) (string, int, error) {
+	log.Printf("Searching for matching ignore pattern for %s in %s\n", ignorePath, ignoreFile)
+	_, err := os.Stat(ignoreFile)
+	if os.IsNotExist(err) {
+		return "", 0, nil
+	}
+
+	if err != nil {
+		return "", 0, err
+	}
+
+	file, err := os.Open(ignoreFile)
+	if err != nil {
+		return "", 0, err
+	}
+	defer file.Close()
+
+	reader := bufio.NewReader(file)
+	lineNumber := 0
+
+	for {
+		pattern := ""
+		isEof := false
+
+		for {
+			linebytes, isprefix, err := reader.ReadLine()
+			pattern = pattern + string(linebytes)
+
+			if isprefix {
+				continue
+			}
+
+			lineNumber++
+
+			if err == io.EOF {
+				isEof = true
+			}
+			break
+		}
+
+		log.Printf("Checking pattern %s in %s\n", pattern, ignoreFile)
+		if pattern == "" || strings.HasPrefix(pattern, "#") {
+			if isEof {
+				break
+			}
+			continue
+		}
+
+		matched := matchesGlob("/"+ignorePath, isDir, pattern)
+		if matched {
+			return pattern, lineNumber, nil
+		}
+
+		if isEof {
+			break
+		}
+	}
+
+	return "", 0, nil
+}
+
+func matchesGlob(path string, isDir bool, pattern string) bool {
```

Comment:
For now, I was going to keep this local because it doesn't quite fit the File.MatchGlob pattern after all. For example, this ones needs to know if the path points to a directory or not for some of the matching logic, but the path may be relative to arbitrary positions within the work directory.